File: 프로그래머스/0/181916. 주사위 게임 3/주사위 게임 3.java
import java.util.*;

class Solution {
    public int solution(int a, int b, int c, int d) {
        int answer = 0;
        Map<Integer, Integer> map = new HashMap<>();
        int[] nums = {a, b, c, d};
        
        for (int num : nums) {
            map.put(num, map.getOrDefault(num, 0) + 1);
        }
        
        // 키 하나 => 하나 있는 키 가져오자
        if (map.size() == 1) {
            for (int key : map.keySet()) {
                answer += key * 1111;
            }
            return answer;
        }
        
        // 키 4개
        if (map.size() == 4) {
            answer = 6;
            for (int key : map.keySet()) {
                answer = Math.min(answer, key);
            }
            
            return answer;
        }
        
        // 키 3개
        if (map.size() == 3) {
            // 키 하나만 있는거 담자
            answer = 1;
            
            for (int key : map.keySet()) {
                if (map.get(key) == 1) {
                    answer *= key;
                }
            }
            
            return answer;
        }
        
        // 키 2개
        // 3, 1 조합
        if (map.containsValue(3)) {
            for (int key : map.keySet()) {
                if (map.get(key) == 3) {
                    answer += 10 * key;
                } else {
                    answer += key;
                }
            }
            
            return answer * answer;
        }
        
        // 2, 2 조합
        List<Integer> list = new ArrayList<>();
        for (int key : map.keySet()) {
            list.add(key);
        }
        
        int one = list.get(0);
        int two = list.get(1);
        
        return (one + two) * Math.abs(one - two);
    }
}

File: 프로그래머스/0/181916. 주사위 게임 3/주사위 게임 3.cpp
#include <bits/stdc++.h>

using namespace std;

int solution(int a, int b, int c, int d) {
    vector<int> nums = {a, b, c, d};
    
    // 오름차순 정렬
    sort(nums.begin(), nums.end());
    
    // 모든 숫자가 같다
    if (nums[0] == nums[3]) return 1111 * nums[0];
    
    // 같은 숫자가 세개인 것이 있다
    if (nums[0] == nums[2] || nums[1] == nums[3]) {
        int answer;
        
        if (nums[0] == nums[2]) answer = 10 * nums[0] + nums[3];
        else answer = 10 * nums[1] + nums[0];
        
        return answer * answer;
    }

    // 같은 숫자가 두 개인 것이 두 케이스 있다.
    if (nums[0] == nums[1] && nums[2] == nums[3]) return (nums[0] + nums[2]) * (abs(nums[0] - nums[2]));
    
    // 같은 숫자가 두 개인 것이 한 케이스 있다.
    if (nums[0] == nums[1] || nums[1] == nums[2] || nums[2] == nums[3]) {
        if (nums[0] == nums[1]) return nums[2] * nums[3];
        else if (nums[1] == nums[2]) return nums[0] * nums[3]; 
        else return nums[0] * nums[1];
    }
    
    // 모든 숫자가 다르다
    return *min_element(nums.begin(), nums.end());
}